std/encoding.d

```diff
@@ -3380,3 +3380,145 @@ version(unittest)
+*/
+enum BOM {
+    utf32be   = 0,  /// [0x00, 0x00, 0xFE, 0xFF]
+    utf32le   = 1,  /// [0xFF, 0xFE, 0x00, 0x00]
+    utf7      = 2,  /*  [0x2B, 0x2F, 0x76, 0x38]
+                        [0x2B, 0x2F, 0x76, 0x39],
+                        [0x2B, 0x2F, 0x76, 0x2B],
+                        [0x2B, 0x2F, 0x76, 0x2F],
+                        [0x2B, 0x2F, 0x76, 0x38, 0x2D]
+                    */
+    utf1      = 7,  /// [0xF7, 0x64, 0x4C]
+    utfebcdic = 8,  /// [0xDD, 0x73, 0x66, 0x73]
+    scsu      = 9,  /// [0x0E, 0xFE, 0xFF]
+    bocu1     = 10, /// [0xFB, 0xEE, 0x28]
+    gb18030   = 11, /// [0x84, 0x31, 0x95, 0x33]
+    utf8      = 12, /// [0xEF, 0xBB, 0xBF]
+    utf16be   = 13, /// [0xFE, 0xFF]
+    utf16le   = 14, /// [0xFF, 0xFE]
+    invalid   = 15  /// no BOM was found
+}
+
+/** Mapping of a byte sequence to $(B Byte Order Mark (BOM))
+*/
+enum bomEntries = tuple(
+    [
+        BOM.utf32be,
+        BOM.utf32le,
+        BOM.utf7,
+        BOM.utf7,
+        BOM.utf7,
+        BOM.utf7,
+        BOM.utf7,
+        BOM.utf1,
+        BOM.utfebcdic,
+        BOM.scsu,
+        BOM.bocu1,
+        BOM.gb18030,
+        BOM.utf8,
+        BOM.utf16be,
+        BOM.utf16le,
+        BOM.invalid
+    ],
+    tuple(
+        to!(ubyte[])([0x00, 0x00, 0xFE, 0xFF]),
+        to!(ubyte[])([0xFF, 0xFE, 0x00, 0x00]),
+        to!(ubyte[])([0x2B, 0x2F, 0x76, 0x39]),
+        to!(ubyte[])([0x2B, 0x2F, 0x76, 0x2B]),
+        to!(ubyte[])([0x2B, 0x2F, 0x76, 0x2F]),
+        to!(ubyte[])([0x2B, 0x2F, 0x76, 0x38, 0x2D]),
+        to!(ubyte[])([0x2B, 0x2F, 0x76, 0x38]),
+        to!(ubyte[])([0xF7, 0x64, 0x4C]),
+        to!(ubyte[])([0xDD, 0x73, 0x66, 0x73]),
+        to!(ubyte[])([0x0E, 0xFE, 0xFF]),
+        to!(ubyte[])([0xFB, 0xEE, 0x28]),
+        to!(ubyte[])([0x84, 0x31, 0x95, 0x33]),
+        to!(ubyte[])([0xEF, 0xBB, 0xBF]),
+        to!(ubyte[])([0xFE, 0xFF]),
+        to!(ubyte[])([0xFF, 0xFE]),
+        new ubyte[0]
+    )
+);
+
+unittest
+{
+    static assert(bomEntries[0].length == bomEntries[1].length);
+}
+
+/** Returns the index of the $(D BOM) into $(D bomEntries).
+
+Params:
+    input = The sequence to check for the $(D BOM)
+
+Returns:
+    the index of the found $(D BOM) or $(D bomEntries[1].length - 1)
+*/
+size_t getBOMTableIndex(Range)(Range input)
+        if(isInputRange!Range)
+{
+    import std.algorithm.searching : startsWith;
+    foreach (idx, it; bomEntries[1])
+    {
+        if (startsWith(input, it))
+        {
+            return idx;
+        }
+    }
+
+    return bomEntries[1].length;
+}
+
+/** Returns the $(D BOM) for a given $(D input).
+If no $(D BOM) can be matched the $(D input) $(D BOM.invalid) is returned.
+
+Params:
+    input = The sequence to check for the $(D BOM)
+
+Returns:
+    the found $(D BOM) or $(D BOM.invalid).
+*/
+BOM getBOM(Range)(Range input)
+        if(isInputRange!Range)
+{
+    immutable idx = getBOMTableIndex(input);
+    return bomEntries[0][idx];
```

Comment:
This isn't as useful as could be. For example, if it returns utf7, how do you know what sequence matched? Why not make the table an array of tuples, with the first element being the enum and the second being the matched bytes?